pkg/controller/statefulset/stateful_set_control_test.go

```diff
@@ -728,6 +729,293 @@ func TestStatefulSetControl_getSetRevisions(t *testing.T) {
 		})
 }
 
+func setupPodManagementPolicy(podManagementPolicy apps.PodManagementPolicyType, set *apps.StatefulSet) *apps.StatefulSet {
+	set.Spec.PodManagementPolicy = podManagementPolicy
+	return set
+}
+
+func TestStatefulSetControlRollingUpdateWithMaxUnavailable(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.MaxUnavailableStatefulSet, true)()
+
+	testCases := []apps.PodManagementPolicyType{
+		apps.OrderedReadyPodManagement,
+		apps.ParallelPodManagement,
+	}
+	for _, tc := range testCases {
+		// Setup the statefulSet controller
+		totalPods := 6
+		var partition int32 = 3
+		var maxUnavailable = intstr.FromInt(2)
+		set := setupPodManagementPolicy(tc, newStatefulSet(totalPods))
+		set.Spec.UpdateStrategy = apps.StatefulSetUpdateStrategy{
+			Type: apps.RollingUpdateStatefulSetStrategyType,
+			RollingUpdate: func() *apps.RollingUpdateStatefulSetStrategy {
+				return &apps.RollingUpdateStatefulSetStrategy{
+					Partition:      &partition,
+					MaxUnavailable: &maxUnavailable,
+				}
+			}(),
+		}
+
+		client := fake.NewSimpleClientset()
+		spc, _, ssc := setupController(client)
+		if err := scaleUpStatefulSetControl(set, ssc, spc, assertBurstInvariants); err != nil {
+			t.Fatal(err)
+		}
+		set, err := spc.setsLister.StatefulSets(set.Namespace).Get(set.Name)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// Change the image to trigger an update
+		set.Spec.Template.Spec.Containers[0].Image = "foo"
+
+		selector, err := metav1.LabelSelectorAsSelector(set.Spec.Selector)
+		if err != nil {
+			t.Fatal(err)
+		}
+		originalPods, err := spc.podsLister.Pods(set.Namespace).List(selector)
+		if err != nil {
+			t.Fatal(err)
+		}
+		sort.Sort(ascendingOrdinal(originalPods))
+
+		// since maxUnavailable is 2, update pods 4 and 5, this will delete the pod 4 and 5,
+		if _, err = ssc.UpdateStatefulSet(context.TODO(), set, originalPods); err != nil {
+			t.Fatal(err)
+		}
+		pods, err := spc.podsLister.Pods(set.Namespace).List(selector)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		sort.Sort(ascendingOrdinal(pods))
+
+		// expected number of pod is 0,1,2,3
+		if !reflect.DeepEqual(pods, originalPods[:4]) {
+			t.Fatalf("Expected pods %v, got pods %v", originalPods[:4], pods)
+		}
+
+		// create new pods
+		if _, err = ssc.UpdateStatefulSet(context.TODO(), set, pods); err != nil {
+			t.Fatal(err)
+		}
+		pods, err = spc.podsLister.Pods(set.Namespace).List(selector)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if tc == apps.OrderedReadyPodManagement {
+			// only one pod gets created at a time due to OrderedReady
```

Comment:
This "loop with huge nested if" is not a good way to do a table driven test BTW. Better to extract logic to e.g. functions which are in the test cases and can just be run. I'm not sure if I'm asking for a change or not because I'm still trying to understand the test.

Reply:
in my original PR, there were two different tests , one for OrderedReadyPodManagement and one for ParallelPodManagement. Since there was lot of duplication between two, i changed it to single test